### PR TITLE
Propagate compilation failures

### PR DESCRIPTION
**Pull request by Flakebi**
Compilation errors were ignored previously, now they fail compilation.
The `InlineAsmDiagHandler` is copied from llc.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit be8efd08a0a4d142c3807fb92d3174f3fec96705
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 95494513896a18924512ef426834e4c49967da56</li>
<li>SPVGEN: dev 6e35eab398b519b8056a13a57dafc6c64b3df13a</li>
<li>XGL: dev 181630cb6bf68773d2840353c7dbe69530bb1819</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fef6a69e1cc922638cd248e8d73670df38fa63a1</li>
</ul></details>

<details><summary>CTS tests (Failed: 2/177789)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 27608/59232 (46.6%)</li>
<li>Failed: 0/59232 (0.0%)</li>
<li>Skipped: 31624/59232 (53.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 27606/59232 (46.6%)</li>
<li><details><summary>Failed: 1/59232 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.vote.compute.subgroupallequal_i8vec3
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 31625/59232 (53.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 26812/59325 (45.2%)</li>
<li><details><summary>Failed: 1/59325 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.creation_feedback.graphics_tests.vertex_stage_fragment_stage_no_cache
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 32512/59325 (54.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

File: llpc/context/llpcCompiler.cpp [hunk 1]
```diff
@@ -211,10 +211,28 @@ static void fatalErrorHandler(void *userData, const std::string &reason, bool ge
 #endif
 }
 
+static void InlineAsmDiagHandler(const SMDiagnostic &SMD, void *Context, unsigned LocCookie) {
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:214]: Please add function header comments and pay attention to the coding style
SMD -> sourceMgrDiag
Context -> context
LocCookie -> locCookie

File: llpc/context/llpcCompiler.cpp [hunk 2]
```diff
@@ -211,10 +211,28 @@ static void fatalErrorHandler(void *userData, const std::string &reason, bool ge
...
+  bool *HasError = static_cast<bool *>(Context);
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:215]: HasError -> hasError

File: llpc/context/llpcCompiler.cpp [hunk 3]
```diff
@@ -211,10 +211,28 @@ static void fatalErrorHandler(void *userData, const std::string &reason, bool ge
...
+
+  // For testing purposes, we print the LocCookie here.
+  if (LocCookie)
+    WithColor::note() << "!srcloc = " << LocCookie << "\n";
+}
+
 // =====================================================================================================================
 // Handler for diagnosis in pass run, derived from the standard one.
 class LlpcDiagnosticHandler : public DiagnosticHandler {
+  bool *HasError;
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:229]: Please add it to a private section with explicit "private". And a private section always follows a public section (not in the beginning of a class).
HasError -> m_hasError

File: llpc/context/llpcCompiler.cpp [hunk 4]
```diff
@@ -211,10 +211,28 @@ static void fatalErrorHandler(void *userData, const std::string &reason, bool ge
...
+
+public:
+  LlpcDiagnosticHandler(bool *HasErrorPtr) : HasError(HasErrorPtr) {}
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:232]: HasErrorPtr -> hasError

File: llpc/context/llpcCompiler.cpp [hunk 5]
```diff
@@ -561,9 +579,11 @@ Result Compiler::BuildShaderModule(const ShaderModuleBuildInfo *shaderInfo, Shad
           result = m_shaderCache->retrieveShader(hEntry, &cacheData, &allocSize);
       }
       if (cacheResult != Result::Success && cacheEntryState != ShaderEntryState::Ready) {
+        bool HasError = false;
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:582]: HasError -> hasError

File: llpc/context/llpcCompiler.cpp [hunk 6]
```diff
@@ -902,16 +927,27 @@ Result Compiler::buildPipelineWithRelocatableElf(Context *context, ArrayRef<cons
...
-    // Return the first relocatable shader, since we can only return one anyway.
-    for (unsigned stage = 0; stage < ShaderStageNativeStageCount; ++stage) {
-      if (elf[stage].empty())
-        continue;
-      *pipelineElf = elf[stage];
-      break;
+  if (result == Result::Success) {
+    if (!isUnlinkedPipeline) {
+      // Link the relocatable shaders into a single pipeline elf file.
+      bool HasError = false;
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:933]: -> hasError

File: llpc/context/llpcCompiler.cpp [hunk 7]
```diff
@@ -1083,7 +1119,9 @@ Result Compiler::buildPipelineInternal(Context *context, ArrayRef<const Pipeline
   TimerProfiler timerProfiler(context->getPipelineHashCode(), "LLPC", TimerProfiler::PipelineTimerEnableMask);
   bool buildingRelocatableElf = context->getPipelineContext()->isUnlinked();
 
-  context->setDiagnosticHandler(std::make_unique<LlpcDiagnosticHandler>());
+  bool HasError = false;
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:1122]: -> hasError

File: llpc/context/llpcCompiler.cpp [hunk 8]
```diff
@@ -211,10 +211,28 @@ static void fatalErrorHandler(void *userData, const std::string &reason, bool ge
...
 }
 
+static void InlineAsmDiagHandler(const SMDiagnostic &SMD, void *Context, unsigned LocCookie) {
+  bool *HasError = static_cast<bool *>(Context);
+  if (SMD.getKind() == SourceMgr::DK_Error)
+    *HasError = true;
+
+  SMD.print(nullptr, errs());
+
+  // For testing purposes, we print the LocCookie here.
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:221]: LocCookie -> locCookie

**Comment by Flakebi** [comment]
Thanks, fixed the formatting.

**Review by amdrexu — APPROVED** [review]

**Comment by amdrexu** [comment]
Thank you for fixing the formatting issues. Looks good now.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
@Flakebi  please see internal discussion.

**Pull request merged**